Add `done` event

daniel-bryant opened the pull request:
See my original issue https://github.com/binaryminds/react-native-sse/issues/56 for some background.

Currently the library only supports `close` events, which fire when the connection is closed from the **client** side. An additional event for tracking **server** side closing of the connection would be useful. This PR adds a `done` event to do that. Luckily most of the plumbing is already there, so it was easy to hook up the new event. As far as naming goes, I just matched the state name like most other events seem to do. It still might be a little confusing, so I added some text to the README.

daniel-bryant committed:
Add done event

daniel-bryant committed:
Update README.md

EmilJunker commented on README.md:
```diff
@@ -51,11 +51,23 @@ es.addEventListener("error", (event) => {
+### Done vs Close
+
+`done` events will fire when server closes the connection.
+Reconnections will occur indefinitely, unless this behavior is disabled.
```
I think this should be rephrased slightly to make it more clear. Maybe something like:

By default, the client will automatically reconnect when this happens. You can disable the reconnections by setting the `pollingInterval` option to 0".

daniel-bryant replied:
Sounds good to me. Updated.

daniel-bryant committed:
Add more clarity to README 'Done vs Close'

EmilJunker commented on src/EventSource.js:
```diff
@@ -115,6 +116,7 @@ class EventSource {
           this._handleEvent(xhr.responseText || '');
 
           if (xhr.readyState === XMLHttpRequest.DONE) {
+            this.dispatch('done', { type: 'done' });
```
I would strongly recommend to dispatch the `done` event _after_ calling `_pollAgain`. This would allow users to cancel the reconnection by calling `.close()` from the `done` handler.

daniel-bryant replied:
Good call. I just updated and tested that it works as expected.

daniel-bryant committed:
Dispatch 'done' event after '_pollAgain'.

This allows 'close' called from a 'done' handler to cancel the upcoming
reconnection. Otherwise '_pollTimer' will get reset after the handler.

EmilJunker commented:
LGTM 👍

daniel-bryant commented:
@quentez that is great to hear, happy to help!

amin-hashthink commented:
it's not merged into the master branch (not available in the latest release, 1.2.1). how can I use this fix @daniel-bryant ?

amin-hashthink commented:
> it's not merged into the master branch (not available in the latest release, 1.2.1). how can I use this fix @daniel-bryant ?

was more of a git/ package management question actually!
for anyone else with the same question, that can be done with a command like this which mentions the git repo and the exact commit (I chose the the latest commit @daniel-bryant added):
`bun add react-native-sse@git+https://github.com/binaryminds/react-native-sse.git#6b59efd7b9f410e30bf029547db50581f9b810ae`

but pls let me know if there's a better way!
